kingpin/actors/rightscale/test/test_base.py

```diff
@@ -225,10 +238,86 @@ def test_add_resource_tags(self):
+            tornado_value(None)
         ]
         yield self.actor._delete_resource_tags(
             resource=resource, tags=['a', 'b'])
         self.client_mock.delete_resource_tags.assert_has_calls([
             mock.call(resource, ['a', 'b'])
         ])
+
+    @testing.gen_test
+    def test_ensure_tags_with_single_string(self):
+        mci = mock.MagicMock(name='mci')
+        mci.href = '/test'
+        self.client_mock.get_resource_tags = mock.MagicMock(name='get')
+        self.client_mock.get_resource_tags.side_effect = [
+            tornado_value(['tag1', 'tag2'])
+        ]
+
+        self.client_mock.add_resource_tags = mock.MagicMock(name='add')
+        self.client_mock.add_resource_tags.side_effect = [
+            tornado_value(None)
+        ]
+
+        self.client_mock.delete_resource_tags = mock.MagicMock(name='delete')
+        self.client_mock.delete_resource_tags.side_effect = [
+            tornado_value(None)
+        ]
+
+        yield self.actor._ensure_tags(mci, 'tag')
+
+        self.client_mock.add_resource_tags.assert_has_calls([
+            mock.call(mci, ['tag'])
+        ])
+        self.client_mock.delete_resource_tags.assert_has_calls([
+            mock.call(mci, ['tag1', 'tag2'])
+        ])
+
+    @testing.gen_test
+    def test_ensure_tags_with_mocked_mci(self):
+        mci = mock.MagicMock(name='mci')
+        mci.href = None
+        self.client_mock.get_resource_tags = mock.MagicMock(name='get')
+        self.client_mock.get_resource_tags.side_effect = [
+            tornado_value(['tag1', 'tag2'])
+        ]
+
+        self.client_mock.add_resource_tags = mock.MagicMock(name='add')
+        self.client_mock.add_resource_tags.side_effect = [
+            tornado_value(None)
+        ]
+
+        self.client_mock.delete_resource_tags = mock.MagicMock(name='delete')
+        self.client_mock.delete_resource_tags.side_effect = [
+            tornado_value(None)
+        ]
+
+        yield self.actor._ensure_tags(mci, 'tag')
+
+        self.client_mock.add_resource_tags.assert_has_calls([
+            mock.call(mci, ['tag'])
+        ])
+        self.assertFalse(self.client_mock.delete_resource_tags.called)
+
+
+class TestEnsurableRightScaleBaseActor(testing.AsyncTestCase):
+
+    def setUp(self, *args, **kwargs):
+        super(TestEnsurableRightScaleBaseActor, self).setUp()
+        base.TOKEN = 'unittest'
+
+        # Create the actor
+        self.actor = FakeEnsurableRightScaleBaseActor(
+            'Copy UnitTestArray to NewUnitArray', {})
+
+        # Patch the actor so that we use the client mock
+        self.client_mock = mock.MagicMock()
+        self.actor._client = self.client_mock
+
+    @testing.gen_test
+    def test_init_without_environment_creds(self):
+        # Un-set the token and make sure the init fails
+        base.TOKEN = None
+        with self.assertRaises(exceptions.InvalidCredentials):
+            FakeEnsurableRightScaleBaseActor('Unit Test Action', {})
```

Comment:
I don't understand why this works... If `base.TOKEN` is None then the `__init__` of `RightScaleBaseActor` should raise the exception. But for some reason inside of the base actor the token is unittest, and the Ensurable base actor has token as None. I think outside of unit tests this situation will never happen.